crates/bonsaidb-macros/src/lib.rs

```diff
@@ -10,63 +10,159 @@
+    #[attribute(
+        missing = r#"You need to specify the collection authority via `#[collection(authority = "authority")]`"#
+    )]
+    authority: String,
+    #[attribute(
+        missing = r#"You need to specify the collection name via `#[collection(name = "name")]`"#
+    )]
+    name: String,
+    #[attribute(default)]
+    #[attribute(expected = r#"Specify the `views` like so: `view = [SomeView, AnotherView]`"#)]
+    views: Vec<Type>,
+    #[attribute(expected = r#"Specify the `serialization` like so: `serialization = Format` or `serialization = None` to disable deriving it"#)]
+    serialization: Option<Path>,
+}
+
+/// Derives the `bonsaidb::core::schema::Collection` trait.
 #[proc_macro_error]
-#[proc_macro_derive(Action)]
-pub fn permissions_action_derive(input: TokenStream) -> TokenStream {
-    let input = parse_macro_input!(input as DeriveInput);
-
-    let name = input.ident;
-
-    let mut fields = Vec::new();
-    match input.data {
-        Data::Enum(data) => {
-            for variant in data.variants.iter() {
-                let ident = variant.ident.clone();
-                let ident_as_string = ident.to_string();
-                match variant.fields.len() {
-                    0 => {
-                        fields.push(quote! { Self::#ident => ActionName(vec![::std::borrow::Cow::Borrowed(#ident_as_string)]) });
-                    }
-                    1 => {
-                        fields.push(quote! {
-                            Self::#ident(subaction) => {
-                                let mut name = Action::name(subaction);
-                                name.0.insert(0, ::std::borrow::Cow::Borrowed(#ident_as_string));
-                                name
-                            }
-                        });
-                    }
-                    _ => {
-                        abort!(
-                            variant.ident,
-                            "For derive(Action), all enum variants may have at most 1 field"
-                        )
-                    }
+/// `#[collection(authority = "Authority", name = "Name", views = [a, b, c])]`
+#[proc_macro_derive(Collection, attributes(collection))]
+pub fn collection_derive(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let DeriveInput {
+        attrs,
+        ident,
+        generics,
+        ..
+    } = parse_macro_input!(input as DeriveInput);
+
+    let CollectionAttribute {
+        authority,
+        name,
+        views,
+        serialization,
+    } = CollectionAttribute::from_attributes(attrs).unwrap_or_abort();
+
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    let serialization = match serialization {
+        Some(serialization) if serialization.is_ident("None") => TokenStream::new(),
+        Some(serialization) => quote! {
+            impl #impl_generics ::bonsaidb::core::schema::SerializedCollection for #ident #ty_generics #where_clause {
```

Comment:
We need to allow someone to specify an alias for where bonsaidb_core is located. I'm good with the default, but someone can import `bonsaidb-server` directly or `bonsaidb-client`, and the import path [will become](https://github.com/khonsulabs/bonsaidb/blob/main/crates/bonsaidb-local/src/lib.rs#L36) `bonsaidb_server::core` or `bonsaidb_client::core`.

This will also allow us to use it internally in our own testing. `bonsaidb_core::test_util` is where macros that define a test suite are defined, and those macros are instantiated in each of the crates and run with a customized harness in each crate. Without the ability to specify where `bonsaidb_core` is located, this macro can't include tests with the macros. This is [how actionable is used this way in bonsaidb_server](https://github.com/khonsulabs/bonsaidb/blob/main/crates/bonsaidb-server/src/server.rs#L1037), although that usage isn't not strictly needed since the dependency could just be added directly.

Reply:
Ok, so there are two ways, to do this, we could have a parameter that you specify for each call, or we use features for the different usages, so you only have to specify it in your Cargo.toml.

I mean ideal would be if macros could depend on core directly and have a "__private" path to it, but that would be a circular dependency :D.

Reply:
We could also use this: https://crates.io/crates/proc-macro-crate to support it automaticaly choose whatever is availible.

It feals a bit hacky but AFAICT it seams to work without problems.

It could be put behind a feature that can be disabled.


So my prefered way of doing this would be:

- [ ] add features for all the standard paths (e.g. `bonsaidb`, `bonsaidb_server` and `bonsaidb_client`) so that those crates can use this feature to indicate to `core` it's path.
- [ ] support the use of [`proc-macro-crate`](https://lib.rs/crates/proc-macro-crate) to find the correct crate (this would be off when any of the explicit crate-path features is enabled)
- [ ] support defining the path via `crate = ::bonsaidb_path::core` via attribute.


What do you think about this @ecton ?

Reply:
I don't love the proc-macro-crate idea. While it might solve some common use cases, it feels too magic and seems as if it will break if you're including BonsaiDb through a second crate. One thing I love about my `actionable` is that you can avoid listing the dependency if you're willing to specify `actionable = bonsaidb::core::actionable`. If I included BonsaiDb in a root crate of my workspace and re-export it to the rest of my crates, it would be nice to not have to add the dependency to each of the other crates to make it work.

I don't believe there's a way to make the featuer-flag version work, because feature flags are additive. We add an extra flag to the child crates `included-from-omnibus`, but I don't see a way to use a feature flag as an inverse -- ie, only pass a feature flag from bonsaidb-local to bonsaidb-core if `included-from-omnibus` *isn't* set. If you suggest a path where the flags can help make the default use case work for directly including those crates, let's try!

I think the attribute rename is the cleanest solution for now, and it doesn't involve anything hacky . I'd prefer the argument to be named `core` instead of `crate`, however, as I don't believe it's obvious *which* crate should be pointed to if the name is just `crate`.

Reply:
I stand corrected via Discord. Let's do it. Idea for feature flags:

- add `included-from-omnibus`, `included-from-client`, `included-from-server`, `included-from-local`, to bonsaidb_core. Activate the flag from each Cargo.toml
- Use this precedence for resolution:
  - If `core` argument is specified, use it.
  - Try to detect the crate using proc-macro-crate
  - Fallback to using feature flags in this precedence order:
    - `included-from-omnibus`: `bonsaidb::core`
    - `included-from-server`: `bonsaidb_server::core`
    - `included-from-local`: `bonsaidb_local::core`
    - `included-from-client`: `bonsaidb_client::core`